### DISCUSSION TITLE
recognition_predictor has poor results on cropped images

### DISCUSSION THREAD

**Discussion by akarshsoman** [mindee/doctr#893] [General]
i have an object detection model train to extract specific text from check image and when i pass it to recognizer it performs poorly , is there anyway to improve it 
![AJCPB3077F1623421014176427](https://user-images.githubusercontent.com/25591311/163812160-80133f0a-2758-4a00-aa50-196dc3dd1dc5.jpeg)
the code used
```python
from doctr.models import ocr_predictor
from doctr.io import DocumentFile

predictor = ocr_predictor(reco_arch='sar_resnet31', pretrained=True)
img = DocumentFile.from_images('/content/AJCPB3077F1623421014176427.jpeg')
res1 = predictor(img)
res2 = predictor.reco_predictor(multi_img_doc)
res2
```
 it returns output :KbKF

**Comment by akarshsoman** [comment]
@felixdittrich92 and @fg-mindee plz can u help

**Comment by felixdittrich92** [comment]
Hi @akarshsoman 👋 ,
My apologies for the late response. I think in your case we have 2 problems:

- detection was only trained on whole document detection not on word line crops or similar 
- recognition models takes only word crops no line crops (to provide only the word crops to the recognizer would solve the problem ^^)

Does the information you need follow any pattern like each starts with IFS CODE: or similar?

I see currently 2 options if you have enough data you can train your own detection model for your line crops or you use the OCR on the whole document and extract your Information with NLP

@charlesmindee correct me if i'm wrong or if you have any other idea :)

**Reply by akarshsoman** [reply]
thankyou @felixdittrich92 for your valuable reply.
can i use east or craft detector and then pass it to the detection model

**Reply by felixdittrich92** [reply]
Hi @akarshsoman,

yes of course. For example you could also use `pytesseract ` only for detect the words in your word crop and after this pass the word crops to the recognition predictor.

Let me know if you have further questions or need help :+1: :smiley:

**Reply by akarshsoman** [reply]
i tried pytesseract + recognition model but still its poor

rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
results = pytesseract.image_to_data(rgb, output_type=Output.DICT)
for i in range(0, len(results["text"])):
    # extract the bounding box coordinates of the text region from
    # the current result
    x = results["left"][i]
    y = results["top"][i]
    w = results["width"][i]
    h = results["height"][i]
    img = rgb[y:y+h,x:x+w]
    cv2.imwrite(f'{i}.jpg',img)

from doctr.models import ocr_predictor
from doctr.io import DocumentFile

predictor = ocr_predictor(reco_arch='sar_resnet31', pretrained=True)
img= DocumentFile.from_images('6.jpg')
res1 = predictor(img)
res2 = predictor.reco_predictor(img)
res2

[('~~~~~~~~~~~~฿฿฿฿฿฿฿฿฿฿฿฿฿฿฿฿฿฿฿', 0.5460261702537537)]

this is the output 

while running the code i get the following warnings:




WARNING:tensorflow:Unresolved object in checkpoint: (root).cfg.fpn_layers
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.0.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.0.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.0.layer-2._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.1.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.1.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.1.layer-2._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.1.layer-3._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.2.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.2.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.2.layer-2._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.2.layer-3._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.3.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.3.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.3.layer-2._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.3.layer-3._inbound_nodes
WARNING:tensorflow:A checkpoint was restored (e.g. tf.train.Checkpoint.restore or tf.keras.Model.load_weights) but not all checkpointed values were used. See above for specific issues. Use expect_partial() on the load status object, e.g. tf.train.Checkpoint.restore(...).expect_partial(), to silence these warnings, or use assert_consumed() to make the check explicit. See https://www.tensorflow.org/guide/checkpoint#loading_mechanics for details.
WARNING:tensorflow:Unresolved object in checkpoint: (root).cfg.fpn_layers
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.0.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.0.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.0.layer-2._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.1.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.1.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.1.layer-2._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.1.layer-3._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.2.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.2.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.2.layer-2._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.2.layer-3._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.3.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.3.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.3.layer-2._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.3.layer-3._inbound_nodes
WARNING:tensorflow:A checkpoint was restored (e.g. tf.train.Checkpoint.restore or tf.keras.Model.load_weights) but not all checkpointed values were used. See above for specific issues. Use expect_partial() on the load status object, e.g. tf.train.Checkpoint.restore(...).expect_partial(), to silence these warnings, or use assert_consumed() to make the check explicit. See https://www.tensorflow.org/guide/checkpoint#loading_mechanics for details.
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-4.layer_with_weights-0
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-7.layer_with_weights-0
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-7.layer_with_weights-1
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-8.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-9.axis
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-9.gamma
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-9.beta
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-9.moving_mean
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-9.moving_variance
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-10.layer_with_weights-0
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-10.layer_with_weights-1
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-10.layer_with_weights-2
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-10.layer_with_weights-3
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-10.layer_with_weights-4
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-11.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-12.axis
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-12.gamma
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-12.beta
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-12.moving_mean
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-12.moving_variance
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-13.layer_with_weights-0
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-13.layer_with_weights-1
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-13.layer_with_weights-2
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-14.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-15.axis
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-15.gamma
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-15.beta
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-15.moving_mean
WARNING:tensorflow:Unresolved object in checkpoint: (root).feat_extractor.layer_with_weights-15.moving_variance

**Reply by felixdittrich92** [reply]
@akarshsoman please attach the cropped image i will test it :)

**Reply by akarshsoman** [reply]
![AIRPA6785F172350205956](https://user-images.githubusercontent.com/25591311/164500727-752cb9ab-0afe-4874-9aac-6582d2213fa2.jpeg)

thanks :)

**Reply by felixdittrich92** [reply]
@akarshsoman 
I think you misunderstood this a bit :)
the recognition_predictor takes only word crops so you have to pass a batch of crops in your case crop from IFS / Code / : / PUNB0070900 to the predictor :+1: 
But i will test this hopefully on monday :smiley:

**Reply by akarshsoman** [reply]
@felixdittrich92 i did pass as word crops. i used tesseract for detection of words and passed it into recognition_predictor .
can u share your code so that i know if i have gone somewhere wrong. thanks :)

**Reply by akarshsoman** [reply]
@felixdittrich92 any update :)

**Reply by felixdittrich92** [reply]
@akarshsoman sorry for the delay i have it on track for tomorrow :)

**Reply by felixdittrich92** [reply]
Hi @akarshsoman 
![testi](https://user-images.githubusercontent.com/23098070/165702054-b7bc1611-e941-40cc-8779-d32cddde59da.png)

```
import pytesseract
from pytesseract import Output
import cv2

from doctr.io import DocumentFile
from doctr.models.recognition.predictor import RecognitionPredictor
from doctr.models.preprocessor import PreProcessor
from doctr.models import crnn_vgg16_bn

image = cv2.imread('/home/felix/Desktop/testi.png')
rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
results = pytesseract.image_to_data(rgb, output_type=Output.DICT)
img_list = []
for i in range(0, len(results["text"])):
    if results['text'][i]: # take only results which not empty
        x = results["left"][i]
        y = results["top"][i]
        w = results["width"][i]
        h = results["height"][i]
        img = rgb[y:y+h,x:x+w]
        img_list.append(img)


reco_model = crnn_vgg16_bn(pretrained=True)
reco_predictor = RecognitionPredictor(PreProcessor((32, 128), preserve_aspect_ratio=True, batch_size=32, mean=(0.694, 0.695, 0.693), std=(0.299, 0.296, 0.301)), reco_model)
out = reco_predictor(img_list)
print(out)
```

`[('IFS', 0.9996910691261292), ('Code', 0.870378315448761), (':', 0.9997653961181641), ('PUNB0070900', 0.9378017783164978)]`

I a solution for the moment: do not use sar & master model :sweat: 
I will investigate a bit more but overall it seems that both models are troublesome also on TF side @frgfm @charlesmindee 

EDIT: crnn_vgg16_bn / crnn_mobilenet_v3_small works well (on PyTorch side no problems all pretrained models works)

**Reply by felixdittrich92** [reply]
@akarshsoman 
Overall it would be great if you can open an detailed issue for this especially about the text detection on text line crops you could link to this discussion and maybe attach a note that sar and master are seems to be broken on TF side. :hugs:

**Reply by frgfm** [reply]
@akarshsoman I can confirm that as of now, it isn't suggested to use SAR as we're trying to debug a few things 😓 
Sorry about the late reply, I was off for some time!

**Reply by akarshsoman** [reply]
Thanks for the help @felixdittrich92 and @frgfm

**Reply by felixdittrich92** [reply]
@akarshsoman both models (SAR and MASTER) are available now in the latest release for TF and PT (without pretrained versions)
